src/s2/util/math/exactfloat/bignum_test.cc

```diff
@@ -0,0 +1,1222 @@
+const int kRandomBignumCount = 128;
+
+static std::vector<std::string> GenerateRandomNumberStrings(
+    absl::BitGenRef bitgen, int bits) {
+  std::vector<std::string> numbers;
+  numbers.reserve(kRandomBignumCount);
+
+  for (int i = 0; i < kRandomBignumCount; ++i) {
+    std::string num;
+
+    // Generate approximately `bits` worth of decimal digits
+    int decimal_digits = (bits * 3) / 10;  // log10(2^bits) ≈ bits * 0.301
+
+    // First digit can't be zero
+    absl::StrAppend(&num, absl::StrFormat("%d", absl::Uniform(bitgen, 1, 9)));
+    for (int j = 1; j < decimal_digits; ++j) {
+      num += absl::Uniform(bitgen, '0', '9');
+    }
+
+    numbers.push_back(num);
+  }
+
+  return numbers;
+}
+
+// Basic correctness test to ensure OpenSSL integration is working
+TEST(BignumTest, OpenSSLIntegration) {
+  OpenSSLBignum a(123);
+  OpenSSLBignum b(456);
+  OpenSSLBignum result;
+
+  BN_add(result.get(), a.get(), b.get());
+
+  char* str = BN_bn2dec(result.get());
+  EXPECT_STREQ(str, "579");
+  OPENSSL_free(str);
+}
+
+TEST(BignumTest, ResultsMatch) {
+  // Test that and OpenSSL produce the same results
+  const Bignum w_a(12345);
+  const Bignum w_b(67890);
+  Bignum w_result = w_a + w_b;
+
+  const OpenSSLBignum ssl_a(12345);
+  const OpenSSLBignum ssl_b(67890);
+  OpenSSLBignum ssl_result;
+  BN_add(ssl_result.get(), ssl_a.get(), ssl_b.get());
+
+  char* ssl_str = BN_bn2dec(ssl_result.get());
+  std::string w_str = absl::StrFormat("%v", w_result);
+
+  EXPECT_EQ(w_str, std::string(ssl_str));
+  OPENSSL_free(ssl_str);
+}
+
+// Different number sizes for benchmarking.
+enum class NumberSizeClass : uint32_t {
+  kSmall = 64,
+  kMedium = 256,
+  kLarge = 1024,
+  kHuge = 4096,
+  kMega = 18000
+};
+
+std::vector<std::string> RandomNumberStrings(absl::BitGenRef bitgen,
+                                             NumberSizeClass size_class) {
+  return GenerateRandomNumberStrings(bitgen, static_cast<int>(size_class));
+}
+
+class VsOpenSSLTest
+    : public TestWithParam<std::pair<NumberSizeClass, NumberSizeClass>> {
+ protected:
+  // Returns a vector of pairs of numbers (as decimal strings) based on the
+  // parameters that the test suite was created with.
+  //
+  // E.g. If the test suite was instantiate with (kSmall, kHuge) as the number
+  // classes, then this will return a small value on the left and a huge value
+  // on the right. (kHuge, kSmall) would return the opposite.
+  std::vector<std::pair<std::string, std::string>> Numbers() {
+    auto numbers0 = RandomNumberStrings(bitgen_, GetParam().first);
+    auto numbers1 = RandomNumberStrings(bitgen_, GetParam().second);
+    ABSL_CHECK_EQ(numbers0.size(), numbers1.size());
+
+    std::vector<std::pair<std::string, std::string>> numbers;
+    numbers.reserve(numbers0.size());
+
+    for (size_t i = 0; i < numbers0.size(); ++i) {
+      numbers.emplace_back(numbers0[i], numbers1[i]);
+    }
+    return numbers;
+  }
+
+ private:
+  absl::BitGen bitgen_;
```

Comment:
defining this in the test case makes it easiest to reproduce

https://github.com/google/s2geometry/blob/d4f3b1bff12c06646a5133e3028559fa670a0d1e/src/s2/s2cell_id_test.cc#L236